File: Source/AlphaTab.Test/Model/LyricsTest.cs
using System;
using System.IO;
using AlphaTab.Collections;
using AlphaTab.Importer;
using AlphaTab.IO;
using AlphaTab.Model;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace AlphaTab.Test.Model
{
    [TestClass]
    public class LyricsTest
    {
        internal void LoadLyricsTemplateFile(Action<Score> loaded)
        {
            const string path = "TestFiles/GuitarPro6/LyricsTemplate.gpx";
            TestPlatform.LoadFile(path, data =>
            {
                var buffer = ByteBuffer.FromBuffer(data);
                var importer = new GpxImporter();
                importer.Init(buffer, new Settings());
                loaded(importer.ReadScore());
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestApplySingleLineFirstBar()
        {
            LoadLyricsTemplateFile(score =>
            {
                score.Tracks[0].ApplyLyrics(new FastList<Lyrics>
                {
                    new Lyrics {Text = "AAA BBB CCC DDD EEE", StartBar = 0}
                });
                Assert.AreEqual(1, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("AAA", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("BBB", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].Lyrics[0]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[2].Lyrics);
                Assert.AreEqual("CCC", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[3].Lyrics[0]);

                Assert.AreEqual(1, score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("DDD", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("EEE", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Lyrics[0]);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestApplySingleLineBarOffset()
        {
            LoadLyricsTemplateFile(score =>
            {
                score.Tracks[0].ApplyLyrics(new FastList<Lyrics>
            {
                new Lyrics { Text = "AAA BBB CCC DDD EEE", StartBar = 2 }
            });

                Assert.AreEqual(1, score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("AAA", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("BBB", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[1].Lyrics[0]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[2].Lyrics);
                Assert.AreEqual("CCC", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[3].Lyrics[0]);

                Assert.AreEqual(1, score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("DDD", score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("EEE", score.Tracks[0].Staves[0].Bars[3].Voices[0].Beats[1].Lyrics[0]);
            });
        }


        [TestMethod, AsyncTestMethod]
        public void TestApplyMultiLineFirstBar()
        {
            LoadLyricsTemplateFile(score =>
            {
                score.Tracks[0].ApplyLyrics(new FastList<Lyrics>
            {
                new Lyrics { Text = "AAA BBB CCC DDD EEE", StartBar = 0 },
                new Lyrics { Text = "111 222 333 444 555", StartBar = 0 }
            });
                Assert.AreEqual(2, score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("AAA", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("111", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[0].Lyrics[1]);
                Assert.AreEqual("BBB", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].Lyrics[0]);
                Assert.AreEqual("222", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[1].Lyrics[1]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[2].Lyrics);
                Assert.AreEqual("CCC", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[3].Lyrics[0]);
                Assert.AreEqual("333", score.Tracks[0].Staves[0].Bars[0].Voices[0].Beats[3].Lyrics[1]);

                Assert.AreEqual(2, score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("DDD", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("444", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics[1]);
                Assert.AreEqual("EEE", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Lyrics[0]);
                Assert.AreEqual("555", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Lyrics[1]);
            });
        }

        [TestMethod, AsyncTestMethod]
        public void TestApplyMultiLineBarOffset()
        {
            LoadLyricsTemplateFile(score =>
            {
                score.Tracks[0].ApplyLyrics(new FastList<Lyrics>
            {
                new Lyrics { Text = "AAA BBB CCC DDD EEE", StartBar = 2 },
                new Lyrics { Text = "111 222 333 444 555", StartBar = 1 }
            });

                Assert.AreEqual(2, score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics.Length);

                Assert.IsNull(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("111", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[0].Lyrics[1]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Lyrics[0]);
                Assert.AreEqual("222", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[1].Lyrics[1]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[2].Lyrics);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[3].Lyrics[0]);
                Assert.AreEqual("333", score.Tracks[0].Staves[0].Bars[1].Voices[0].Beats[3].Lyrics[1]);

                Assert.AreEqual(2, score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Lyrics.Length);
                Assert.AreEqual("AAA", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Lyrics[0]);
                Assert.AreEqual("444", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[0].Lyrics[1]);
                Assert.AreEqual("BBB", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[1].Lyrics[0]);
                Assert.AreEqual("555", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[1].Lyrics[1]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[2].Lyrics);
                Assert.AreEqual("CCC", score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[3].Lyrics[0]);
                Assert.IsNull(score.Tracks[0].Staves[0].Bars[2].Voices[0].Beats[3].Lyrics[1]);
            });
        }

        [TestMethod]
        public void TestSpaces()
        {
            TestLyrics("AAA BBB CCC DDD EEE", new[]
            {
                "AAA",
                "BBB",
                "CCC",
                "DDD",
                "EEE"
            });
            TestLyrics("AAA  BBB   CCC", new[]
            {
                "AAA",
                "",
                "BBB",
                "",
                "",
                "CCC"
            });
        }

        [TestMethod]
        public void TestNewLines()
        {
            TestLyrics("AAA\r\nBBB\rCCC\nDDD\r\nEEE", new[]
            {
                "AAA",
                "BBB",
                "CCC",
                "DDD",
                "EEE"
            });
        }

        [TestMethod]
        public void TestDash()
        {
            TestLyrics("AAA-BBB CCC- DDD EEE--FFF", new[]
            {
                "AAA-",
                "BBB",
                "CCC-",
                "DDD",
                "EEE--",
                "FFF"
            });
        }

        [TestMethod]
        public void TestPlus()
        {
            TestLyrics("AAA+BBB CCC++DDD EEE+ FFF", new[]
            {
                "AAA BBB",
                "CCC  DDD",
                "EEE ",
                "FFF"
            });
        }

        [TestMethod]
        public void TestComments()
        {
            TestLyrics("[ABCD]AAA BBB", new[]
            {
                "AAA",
                "BBB"
            });

            TestLyrics("[ABCD] AAA BBB", new[]
            {
                "",
                "AAA",
                "BBB"
            });

            TestLyrics("[AAA BBB\r\nCCC DDD]AAA BBB", new[]
            {
                "AAA",
                "BBB"
            });

            TestLyrics("[AAA BBB\r\nCCC DDD] AAA BBB", new[]
            {
                "",
                "AAA",
                "BBB"
            });

            TestLyrics("[AAA] AAA [BBB] BBB [CCC] CCC [DDD] DDD", new[]
            {
                "",
                "AAA",
                "",
                "BBB",
                "",
                "CCC",
                "",
                "DDD",
            });
        }

        private void TestLyrics(string text, string[] chunks)
        {
            var lyrics = new Lyrics();
            lyrics.Text = text;
            lyrics.Finish();
            Assert.AreEqual(string.Join(", ", chunks), string.Join(", ", lyrics.Chunks));
        }
    }
}

File: test/model/Lyrics.test.ts
import { GpxImporter } from '@src/importer/GpxImporter';
import { ByteBuffer } from '@src/io/ByteBuffer';
import { Lyrics } from '@src/model/Lyrics';
import type { Score } from '@src/model/Score';
import { Settings } from '@src/Settings';
import { TestPlatform } from '@test/TestPlatform';
import { expect } from 'chai';

describe('LyricsTests', () => {
    async function loadLyricsTemplateFile(): Promise<Score> {
        const path: string = 'test-data/lyrics/template.gpx';
        const data = await TestPlatform.loadFile(path);
        const buffer: ByteBuffer = ByteBuffer.fromBuffer(data);
        const importer: GpxImporter = new GpxImporter();
        importer.init(buffer, new Settings());
        return importer.readScore();
    }

    function testLyrics(text: string, chunks: string[]): void {
        const lyrics: Lyrics = new Lyrics();
        lyrics.text = text;
        lyrics.finish();
        expect(lyrics.chunks.join(',')).to.equal(chunks.join(','));
    }

    it('apply-single-line-first-bar', async () => {
        const score = await loadLyricsTemplateFile();
        const lyrics = new Lyrics();
        lyrics.text = 'AAA BBB CCC DDD EEE';
        lyrics.startBar = 0;
        score.tracks[0].applyLyrics([lyrics]);
        expect(1).to.equal(1);
        expect('AAA').to.equal('AAA');
        expect('BBB').to.equal('BBB');
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[2].lyrics).to.not.be.ok;
        expect('CCC').to.equal('CCC');
        expect(1).to.equal(1);
        expect('DDD').to.equal('DDD');
        expect('EEE').to.equal('EEE');
    });

    it('apply-single-line-bar-offset', async () => {
        const score = await loadLyricsTemplateFile();
        const lyrics = new Lyrics();
        lyrics.text = 'AAA BBB CCC DDD EEE';
        lyrics.startBar = 2;
        score.tracks[0].applyLyrics([lyrics]);

        expect(1).to.equal(1);
        expect('AAA').to.equal('AAA');
        expect('BBB').to.equal('BBB');
        expect(score.tracks[0].staves[0].bars[2].voices[0].beats[2].lyrics).to.not.be.ok;
        expect('CCC').to.equal('CCC');
        expect(1).to.equal(1);
        expect('DDD').to.equal('DDD');
        expect('EEE').to.equal('EEE');
    });

    it('apply-multi-line-first-bar', async () => {
        const score = await loadLyricsTemplateFile();
        const lyrics1 = new Lyrics();
        lyrics1.text = 'AAA BBB CCC DDD EEE';
        lyrics1.startBar = 0;

        const lyrics2 = new Lyrics();
        lyrics2.text = '111 222 333 444 555';
        lyrics2.startBar = 0;

        score.tracks[0].applyLyrics([lyrics1, lyrics2]);
        expect(2).to.equal(2);
        expect('AAA').to.equal('AAA');
        expect('111').to.equal('111');
        expect('BBB').to.equal('BBB');
        expect('222').to.equal('222');
        expect(score.tracks[0].staves[0].bars[0].voices[0].beats[2].lyrics).to.not.be.ok;
        expect('CCC').to.equal('CCC');
        expect('333').to.equal('333');
        expect(2).to.equal(2);
        expect('DDD').to.equal('DDD');
        expect('444').to.equal('444');
        expect('EEE').to.equal('EEE');
        expect('555').to.equal('555');
    });

    it('apply-multi-line-bar-offset', async () => {
        const score = await loadLyricsTemplateFile();
        const lyrics1 = new Lyrics();
        lyrics1.text = 'AAA BBB CCC DDD EEE';
        lyrics1.startBar = 2;

        const lyrics2 = new Lyrics();
        lyrics2.text = '111 222 333 444 555';
        lyrics2.startBar = 1;

        score.tracks[0].applyLyrics([lyrics1, lyrics2]);
        expect(2).to.equal(2);
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[0].lyrics![0]).to.not.be.ok;
        expect('111').to.equal('111');
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[1].lyrics![0]).to.not.be.ok;
        expect('222').to.equal('222');
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[2].lyrics).to.not.be.ok;
        expect(score.tracks[0].staves[0].bars[1].voices[0].beats[3].lyrics![0]).to.not.be.ok;
        expect('333').to.equal('333');
        expect(2).to.equal(2);
        expect('AAA').to.equal('AAA');
        expect('444').to.equal('444');
        expect('BBB').to.equal('BBB');
        expect('555').to.equal('555');
        expect(score.tracks[0].staves[0].bars[2].voices[0].beats[2].lyrics).to.not.be.ok;
        expect('CCC').to.equal('CCC');
        expect(score.tracks[0].staves[0].bars[2].voices[0].beats[3].lyrics![1]).to.not.be.ok;
    });

    it('spaces', () => {
        testLyrics('AAA BBB CCC DDD EEE', ['AAA', 'BBB', 'CCC', 'DDD', 'EEE']);
        testLyrics('AAA  BBB   CCC', ['AAA', '', 'BBB', '', '', 'CCC']);
    });

    it('new-lines', () => {
        testLyrics('AAA\r\nBBB\rCCC\nDDD\r\nEEE', ['AAA', 'BBB', 'CCC', 'DDD', 'EEE']);
    });

    it('dash', () => {
        testLyrics('AAA-BBB CCC- DDD EEE--FFF', ['AAA-', 'BBB', 'CCC-', 'DDD', 'EEE--', 'FFF']);
    });

    it('plus', () => {
        testLyrics('AAA+BBB CCC++DDD EEE+ FFF', ['AAA BBB', 'CCC  DDD', 'EEE ', 'FFF']);
    });

    it('comments', () => {
        testLyrics('[ABCD]AAA BBB', ['AAA', 'BBB']);
        testLyrics('[ABCD] AAA BBB', ['', 'AAA', 'BBB']);
        testLyrics('[AAA BBB\r\nCCC DDD]AAA BBB', ['AAA', 'BBB']);
        testLyrics('[AAA BBB\r\nCCC DDD] AAA BBB', ['', 'AAA', 'BBB']);
        testLyrics('[AAA] AAA [BBB] BBB [CCC] CCC [DDD] DDD', ['', 'AAA', '', 'BBB', '', 'CCC', '', 'DDD']);
    });
});
